examples/rust/pipe-ping-pong.rs

```diff
@@ -168,109 +168,164 @@ fn close(libos: &mut LibOS, pipeqd: QDesc) {
     }
 }
 
-//======================================================================================================================
-// server()
-//======================================================================================================================
+// The pipe server.
+pub struct PipeServer {
+    /// Underlying libOS.
+    libos: LibOS,
+    /// Tx pipe qd.
+    tx_pipeqd: Option<QDesc>,
+    /// Rx pipe qd.
+    rx_pipeqd: Option<QDesc>,
+}
 
-fn server(name: &str) -> Result<()> {
-    let libos_name: LibOSName = match LibOSName::from_env() {
-        Ok(libos_name) => libos_name.into(),
-        Err(e) => anyhow::bail!("{:?}", e),
-    };
-    let mut libos: LibOS = match LibOS::new(libos_name) {
-        Ok(libos) => libos,
-        Err(e) => anyhow::bail!("failed to initialize libos: {:?}", e),
-    };
+// Implementation of the pipe server.
+impl PipeServer {
+    pub fn new(libos: LibOS) -> Result<Self> {
+        return Ok(Self {
+            libos,
+            tx_pipeqd: None,
+            rx_pipeqd: None,
+        });
+    }
 
-    // Setup peer.
-    let pipeqd_rx: QDesc = match libos.create_pipe(&format!("{}:rx", name)) {
-        Ok(qd) => qd,
-        Err(e) => anyhow::bail!("failed to open memory queue: {:?}", e),
-    };
+    pub fn run(&mut self, pipe_name: &str) -> Result<()> {
+        // Setup the Rx pipe.
+        self.rx_pipeqd = match self.libos.create_pipe(&format!("{}:rx", pipe_name)) {
+            Ok(qd) => Some(qd),
+            Err(e) => anyhow::bail!("failed to open memory queue: {:?}", e),
+        };
 
-    // Setup peer.
-    let pipeqd_tx: QDesc = match libos.create_pipe(&format!("{}:tx", name)) {
-        Ok(qd) => qd,
-        Err(e) => {
-            // Clean up open pipes on error.
-            close(&mut libos, pipeqd_rx);
-            anyhow::bail!("failed to open memory queue: {:?}", e)
-        },
-    };
+        // Setup the Tx pipe.
+        self.tx_pipeqd = match self.libos.create_pipe(&format!("{}:tx", pipe_name)) {
+            Ok(qd) => Some(qd),
+            Err(e) => {
+                if let Some(rx_pipeqd) = self.rx_pipeqd {
+                    close(&mut self.libos, rx_pipeqd);
+                    self.rx_pipeqd = None;
+                }
+                anyhow::bail!("failed to open memory queue: {:?}", e)
+            },
+        };
```

Comment:
Could you move this setup logic to the constructor?

Reply:
I left this logic in run() because it would be better to create the pipes closest to when we intend to use them.

Reply:
I understand, but I have a different perspective on this and I insist to move this logic to the constructor. Here is the rationale:
1. The point of using associated functions is to rely on a more OO programming pattern.
2. In a such OO pattern, `drop` is used for releasing resources, such as new is used for allocating them.
3. [Having resource acquisition and release tight to the lifetime of an object is a well know programming pattern](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization).
4. Having the object state also initialized in `run()` and to be release in `drop`  does make logic more complex.

Reply:
Done.